### DISCUSSION TITLE
Diagnostic picker doesn't show diagnostics if the filepath for the current file contains a '+'

### DISCUSSION THREAD

**Discussion by JuliusBrueggemann** [helix-editor/helix#7321] [Troubleshooting]
### Summary

If a filepath contains a '+' character, the diagnostic picker will not display diagnostics in that file. The workspace diagnostic picker is unaffected by this. Presumably, other "special" character might also affect the diagnostic picker, even though I did not test with any other characters. 

### Reproduction Steps

Create the file `main.cpp` with the following content in a directory called `c++` (like `~/c++/main.cpp`). clangd will correctly issue a diagnostic for this error (that Helix correctly displays):

![grafik](https://github.com/helix-editor/helix/assets/35491022/06cd4744-2e59-4a34-8324-c6532736feac)

However, when opening the diagnostic picker with Space + d, the diagnostic does not appear:

![grafik](https://github.com/helix-editor/helix/assets/35491022/883a6b53-c6b2-47e1-8119-d6c590ae7f49)

On the other hand, the workspace diagnostic picker (Space + D) does show the diagnostic:

![grafik](https://github.com/helix-editor/helix/assets/35491022/39c5f3bf-c8a8-4ea8-bd0c-1c56b4f5f775)

If `main.cpp` is instead placed in a directory whose path does not contain "special" characters like '+', the diagnostic picker works as expected:

![grafik](https://github.com/helix-editor/helix/assets/35491022/7ac06217-4522-4eb6-a3da-40936d2319b9)

### Platform

Linux

### Terminal Emulator

GNOME Terminal 3.48.1 using VTE 0.72.2 +BIDI +GNUTLS +ICU +SYSTEMD (This is the output from `gnome-terminal --version`)

### Helix Version

helix 23.05 (7f5940be)

**Comment by the-mikedavis** [comment]
I can reproduce this with clangd but not rust-analyzer. clangd percent-encodes the `+`s, so the diagnostics come in with a URI like `file:///tmp/%2B/main.cpp` rather than `file:///tmp/+/main.cpp`. The document diagnostics picker looks up the diagnostics using a URL it derives from the file path and we don't percent-encode the `+`. https://github.com/helix-editor/helix/blob/18e07addfd464844fc18833cfb1d1cd38c683c11/helix-term/src/commands/lsp.rs#L514-L520

I believe we need to normalize the Urls in the `diagnostics` field on `Editor` (https://datatracker.ietf.org/doc/html/rfc3986#section-6.2.2.2), so this looks like another instance of https://github.com/helix-editor/helix/issues/3267

**Reply by pascalkuthe** [reply]
I don't think normalizing is quite enough. We actually need to convert the URLs to paths and cannonicolize them (see https://github.com/helix-editor/helix/pull/3347#issuecomment-1440025499). Diagnostics are the only thing we store as URLs instead of paths (even though the LS alwayr sends URLs)
